akka-http-core/src/main/scala/akka/http/impl/engine/client/OutgoingConnectionBlueprint.scala

```diff
@@ -203,9 +219,25 @@ private[http] object OutgoingConnectionBlueprint {
         }
       }
 
+      lazy val substreamOutHandler = new OutHandler {
+        override def onPull(): Unit = {
+          // upstream pull arrived while streaming entity
+          // we will always pull upon reaching end anyways
+          // so safe to ignore
+        }
+        override def onDownstreamFinish(): Unit = {
+          // if downstream cancels, make sure we also
+          // cancel the entity source
+          if (entitySource != null) entitySource.complete()
+          completeStage()
+        }
+      }
```

Comment:
This should not be necessary: the onPull behavior is already encoded above and the handling of cancellation can also be folded into the normal `onPull` in the same way.